Model/Product/Builder.php

```diff
@@ -397,7 +401,7 @@ private function buildAvailability(Product $product, Store $store)
     {
         $availability = ProductInterface::OUT_OF_STOCK;
         if (!$product->isVisibleInSiteVisibility()
-            || !$this->isAvailabeInStore($product, $store)
+            || (!$this->isAvailableInStore($product, $store) && $this->isInStock($product, $store))
```

Comment:
Why do we need `$this->isInStock($product, $store)` here? I'd at least extract this into a variable since it's used later on as well.

Reply:
Cause `isAvailableInStore` doesn't take into account `outOfStock` items. They can be either. By checking if is `outOfStock` we prevent to mark all of them as invisible when, in fact, they are `outOfStock`